ViewEnvironmentUI/Sources/ViewEnvironmentPropagating.swift

```diff
@@ -105,22 +126,248 @@ extension ViewEnvironmentPropagating {
         return environment
     }
 
+    /// Consumers _must_ call this function when the environment should be re-applied, e.g. in
+    /// `viewWillLayoutSubviews()` for `UIViewController`s and `layoutSubviews()` for `UIView`s.
+    ///
+    /// This will call `apply(environment:)` on the receiver if the node has been flagged for needing update.
+    ///
+    public func applyEnvironmentIfNeeded() {
+        guard needsEnvironmentUpdate else { return }
+
+        needsEnvironmentUpdate = false
+
+        if let observing = self as? ViewEnvironmentObserving {
+            let environment = observing.environment
+            observing.apply(environment: environment)
+        }
+    }
+
     /// Notifies all appropriate descendants that the environment needs update.
     ///
-    /// Ancestor-descendent bindings must be mutually agreed for this method to notify them. If a descendant's 
-    /// `environmentAncestor` is `nil` it will not be notified of needing update.
+    /// Ancestor-descendent bindings must be mutually agreed for this method to notify them. If a descendant's
+    /// `environmentAncestor` is not `self` it will not be notified of needing update.
     ///
     @_spi(ViewEnvironmentWiring)
     public func setNeedsEnvironmentUpdateOnAppropriateDescendants() {
         for descendant in environmentDescendants {
-            // If the descendant's `environmentAncestor` is nil it has opted out of environment updates and is likely
-            // acting as a root for propagation bridging purposes (e.g. from a Workflow ViewEnvironment update).
+            // If the descendant's `environmentAncestor` is not `self` it has opted out of environment updates from this
+            // node. The node is is likely acting as a root for propagation bridging purposes (e.g. from a Workflow
+            // ViewEnvironment update).
             // Avoid updating the descendant if this is the case.
-            guard descendant.environmentAncestor != nil else {
-                continue
+            if descendant.environmentAncestor === self {
+                descendant.setNeedsEnvironmentUpdate()
+            }
+        }
+    }
+
+    /// Adds a `ViewEnvironment` change observation.
+    ///
+    /// The observation will only be active for as long as the returned lifetime is retained or
+    /// `cancel()` is called on it.
+    ///
+    @_spi(ViewEnvironmentWiring)
+    public func addEnvironmentNeedsUpdateObserver(
+        _ onNeedsUpdate: @escaping (ViewEnvironment) -> Void
+    ) -> ViewEnvironmentUpdateObservationLifetime {
+        let object = NSObject()
```

Comment:
Can we make this a specific type (eg, `NeedsEnvironmentUpdateToken` or something) so that it's easier to understand if/when we see these in the memory debugger?

Reply:
I'm still not clear on how this would help / when you'd run into this, but I don't have a strong take. Will update.

- [ ] Use special type for object tokens